### ISSUE TITLE
Manifold volume

### ISSUE THREAD

**Issue by mateuszbaran** [JuliaManifolds/ManifoldMeasures.jl#7]
Currently the `volume` function only depends on the manifold, but I think it should also take measure into account since some manifolds don't have a single universally accepted volume (see for example http://arxiv.org/abs/math-ph/0210033 ).

**Comment by kellertuer** [comment]
I also think that either connection or metric should be given.

**Comment by mateuszbaran** [comment]
It's even more complex, as there are other ways to measure the volume than from the Riemannian metric (see the paper I've linked), and connection is not sufficient as far as I know. I didn't even find a reference for chart-independence of the volume defined through the Riemannian volume form (just a claim on mathoverflow that it's true).

**Comment by sethaxen** [comment]
Here we use "volume" specifically to refer to the total mass of the unnormalized Hausdorff measure, i.e., the quantity that the Hausdorff measure must be divided by to make it a probability measure. The Hausdorff measure assumes a specific metric. When we don't use a `MetricManifold`, the metric assumed is the default metric for the manifold. So I _think_ the use of `volume` with a manifold argument here is valid.

It might be more appropriate though to replace `volume` with `total_mass`, which takes as an argument a measure on a manifold. This would allow us to have the generic `normalize(μ::AbstractMeasure) = inv(total_mass(μ)) * μ`, which maybe should be defined in MeasureTheory.jl to avoid type piracy.

Does this sound right?

**Comment by mateuszbaran** [comment]
Isn't the Hausdorff measure only defined for submanifolds of R^n? It would make sense to me to have an interface for getting the volume of the manifold with respect to some other measure.

**Comment by sethaxen** [comment]
> Isn't the Hausdorff measure only defined for submanifolds of R^n?

I believe it's defined for manifolds with an embedding in a metric space. So e.g. if one embedded the Sphere in R^n with varying radii, this would yield different volumes (surface area).

I need to figure out what I want to do with `Hausdorff`. It's primary use is as a base measure with respect to which log densities are usually defined. But one might want to change to a Lebesgue measure or vice versa, which is where the `log_det_local_metric` term comes in. In practice one would do this when reparametrizing a density defined in coordinates of the embedding to some other coordinate system. My current thinking is to use the atlas machinery in Manifolds for this, but I haven't given it much thought.

> It would make sense to me to have an interface for getting the volume of the manifold with respect to some other measure.

Do you think the proposed `total_mass` would be a good way to do that?

**Comment by mateuszbaran** [comment]
> > 
> 
> I believe it's defined for manifolds with an embedding in a metric space. So e.g. if one embedded the Sphere in R^n with varying radii, this would yield different volumes (surface area).
> 
> I need to figure out what I want to do with `Hausdorff`. It's primary use is as a base measure with respect to which log densities are usually defined. But one might want to change to a Lebesgue measure or vice versa, which is where the `log_det_local_metric` term comes in. In practice one would do this when reparametrizing a density defined in coordinates of the embedding to some other coordinate system. My current thinking is to use the atlas machinery in Manifolds for this, but I haven't given it much thought.

Right, but maybe we'd like to use a Haar measure as a base measure for some group manifolds? It may be different from the Hausdorff measure. `total_mass` nicely solves this issue. I think 'volume' could be `total_mass` with respect to some `default_measure` (like we have `default_retraction`, `default_vector_transport` etc), but not necessarily the Hausdorff measure.

I think atlases should work as a way to change the measure to or from the Lebesgue measure. Just keep in mind that I haven't solved yet the issue of defining things like `log_det_local_metric`, which can be written by hand for a particular atlas (easy when you have a formula), or computed from the metric tensor (not particularly well tested).

**Comment by sethaxen** [comment]
> Right, but maybe we'd like to use a Haar measure as a base measure for some group manifolds? It may be different from the Hausdorff measure. `total_mass` nicely solves this issue. I think 'volume' could be `total_mass` with respect to some `default_measure` (like we have `default_retraction`, `default_vector_transport` etc), but not necessarily the Hausdorff measure.

Great, okay, we're on the same page. Right now there's no notion of a default measure, but that could be added later. I think what @kellertuer needs is something a little more specific, like default sampleable measure, which would work for non-compact manifolds as well. Right now implemented measures just have a base measure, which is useful for defining log densities (and consequently we automatically get rejection sampling). Certainly some measures on groups will have a Haar measure as their base measure.

> I think atlases should work as a way to change the measure to or from the Lebesgue measure. Just keep in mind that I haven't solved yet the issue of defining things like `log_det_local_metric`, which can be written by hand for a particular atlas (easy when you have a formula), or computed from the metric tensor (not particularly well tested).

Thanks for the heads up! I'll hold off on adding that functionality yet and just leave the optional `atlas` field as a placeholder.
